src/main/java/com/github/princesslana/slothbot/middleware/PermissionMiddleware.java

```diff
@@ -0,0 +1,36 @@
+package com.github.princesslana.slothbot.middleware;
+
+import com.github.princesslana.slothbot.Config;
+import com.github.princesslana.slothbot.Discord;
+import com.github.princesslana.slothbot.Moderator;
+import com.google.gson.JsonObject;
+import disparse.discord.smalld.Event;
+import disparse.discord.smalld.permissions.Permission;
+import disparse.discord.smalld.permissions.PermissionUtils;
+import java.util.function.BiFunction;
+
+public class PermissionMiddleware implements BiFunction<Event, String, Boolean> {
+
+  private final Moderator moderator;
+
+  public PermissionMiddleware(Moderator moderator) {
+    this.moderator = moderator;
+  }
+
+  @Override
+  public Boolean apply(Event event, String commandName) {
+    if (Config.getPublicCommands().contains(commandName)) {
+      return true;
+    }
+    if (PermissionUtils.computeAllPerms(event).contains(Permission.ADMINISTRATOR)) {
+      return true;
+    }
+    JsonObject json = event.getJson().getAsJsonObject("d");
+
+    var possibleGuildId = Discord.getGuild(json);
+    var roles = Discord.getRoles(json);
+    return possibleGuildId
+        .map(guildId -> roles.stream().anyMatch(role -> moderator.contains(guildId, role)))
+        .orElse(true);
```

Comment:
Is this saying, if we don't have a guild id allow the command? Feels like this should be false.

Reply:
If we do not have a guildId, it's possible it's a DM command which this middleware probably shouldn't decide if that gets through.  Instead, on the CommandHandler itself we are able to set IncomingScope.CHANNEL and disparse will handle that case for us.

This is good in case you add some command that can work from dm or channel and not just kill the chain.  Such as mentioning the bot or using about or other basic stuff.